pep-0646.rst

```diff
@@ -850,65 +850,81 @@ However, note that slices involving starred expressions are still invalid:
-    | '*' ',' b=param_maybe_default+ c=[kwds] {
-        _PyPegen_star_etc(p, NULL, b, c) }
-    | a=kwds { _PyPegen_star_etc(p, NULL, NULL, a) }
+    star_etc:
+    | '*' param_no_default param_maybe_default* [kwds]
+    | '*' param_no_default_star_annotation param_maybe_default* [kwds]
+    | '*' ',' param_maybe_default+ [kwds]
+    | kwds
 
 Where:
 
 ::
 
-    param_no_default_star_annotation[arg_ty]:
-    | a=param_star_annotation ',' tc=TYPE_COMMENT? { _PyPegen_add_type_comment_to_arg(p, a, tc) }
-    | a=param_star_annotation tc=TYPE_COMMENT? &')' { _PyPegen_add_type_comment_to_arg(p, a, tc) }
+    param_no_default_star_annotation:
+    | param_star_annotation ',' TYPE_COMMENT?
+    | param_star_annotation TYPE_COMMENT? &')'
 
-    param_star_annotation[arg_ty]: a=NAME b=star_annotation { _PyAST_arg(a->v.Name.id, b, NULL, EXTRA) }
+    param_star_annotation: NAME star_annotation
 
-    star_annotation[expr_ty]: ':' a=star_expression { _PyAST_Tuple(_PyPegen_singleton_seq(p, a), Load, EXTRA) }
+    star_annotation: ':' star_expression
 
-This accomplishes the desired outcome (making ``*args: *Ts`` not be a syntax error)
-while matching the behaviour of a starred TypeVarTuple in an index: at runtime,
-``__iter__`` is called on the TypeVarTuple, and a tuple containing the resulting
-unpacked version of the TypeVarTuple is set as the type annotation for ``args``:
+This accomplishes the desired outcome (making ``*args: *Ts`` not be a syntax
+error) while matching the behaviour of star-unpacking in other contexts:
+at runtime, ``__iter__`` is called on the starred object, and a tuple
+containing the items of the resulting iterator is set as the type annotion
+for ``args``. In other words, ``*args: *foo`` is equivalent to
+``*args: tuple(foo)``.
 
 ::
 
     >>> Ts = TypeVarTuple('Ts')
     >>> def foo(*args: *Ts): pass  # Equivalent to `*args: tuple(Ts)`
     >>> foo.__annotations__
-    {'args': (UnpackedTs,)}  # Where UnpackedTs is Ts._unpacked
+    {'args': (*Ts,)}
+    # *Ts is the repr() of Ts._unpacked, an instance of UnpackedTypeVarTuple
 
-Note that the only scenario in which this grammar change allows ``*Ts`` to be used as a direct
-annotation (rather than being wrapped in e.g. ``Tuple[*Ts]``) is ``*args``. Other uses are still
-invalid:
+Note that the only scenario in which this grammar change allows ``*Ts`` to be
+used as a direct annotation (rather than being wrapped in e.g. ``Tuple[*Ts]``)
+is ``*args``. Other uses are still invalid:
 
 ::
 
     x: *Ts                 # Syntax error
     def foo(x: *Ts): pass  # Syntax error
 
+Implications
+''''''''''''
+
+As with the first grammar change, this change also has a number of side effects.
+In particular, the annotation of ``*args`` could be set to a starred object
+other than a ``TypeVarTuple`` - for example, the following nonsensical
+annotation is possible:
+
+::
+
+    >>> l = [1, 2, 3]
+    >>> def bar(*args: *l): pass  # Equivalent to `*args: tuple(l)`
```

Comment:
I'd use a different variable name than `l`, which is easily confused with either `1` or `I` depending on the font.

Reply:
Done.